### PR TITLE
New internal query result payload format: send format from querier if requested + request format from query-frontend if enabled

### PR DESCRIPTION
**Pull request by charleskorn**
#### What this PR does

This PR implements the new internal query result payload format in the querier and query-frontend for the query path (ie. query-frontend -> ruler payloads are still in JSON format).

I've added a flag (`-query-frontend.query-result-payload-format=protobuf`) to enable the new format. This flag causes the query-frotend to send a different `Accept` header with both a new MIME type for our protobuf format and the JSON MIME type, and the querier will only send protobuf responses if the `Accept` header contains the protobuf MIME type.

I've also added an integration test to exercise the new format.

The querier will send native histograms if the query results contain any native histograms, but the query-frontend currently ignores them as [this change to `model.proto`](https://github.com/grafana/mimir/pull/3505/files#diff-2515e0a4916c2504c9c0062bdb6f99d80f64245e21b09a64a3734b49c2c3a216) is not yet merged to `main`.

Note to reviewers: I'd suggest reviewing each commit individually rather than trying to digest the entire change in one go.

#### Which issue(s) this PR fixes or relates to

#4104

#### Checklist

- [x] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/frontend/querymiddleware/codec.go [hunk 1]
```diff
@@ -369,14 +398,17 @@ func (c prometheusCodec) EncodeResponse(ctx context.Context, res Response) (*htt
 		sp.LogFields(otlog.Int("series", len(a.Data.Result)))
 	}
 
+	// TODO: select format based on Accept header
```
Comment (charleskorn) [inline pkg/frontend/querymiddleware/codec.go:401]: Note to reviewers: this is where we'd implement content negotiation to send the protobuf format to rulers.

File: pkg/frontend/querymiddleware/codec_protobuf.go [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+)
+
+type protobufFormat struct{}
+
+func (f protobufFormat) Name() string {
+	return formatProtobuf
+}
+
+func (f protobufFormat) EncodeResponse(resp *PrometheusResponse) ([]byte, error) {
+	panic("not yet implemented")
```
Comment (charleskorn) [inline pkg/frontend/querymiddleware/codec_protobuf.go:21]: Note to reviewers: this method is never called, so this panic is safe. (We only ever use the JSON format in `prometheusCodec.EncodeResponse()`.)

File: pkg/frontend/querymiddleware/codec.go [hunk 3]
```diff
@@ -315,11 +317,16 @@ func (prometheusCodec) EncodeRequest(ctx context.Context, r Request) (*http.Requ
 		Header:     http.Header{},
 	}
 
+	if c.requestProtobufFormat {
```
Comment (56quarters) [inline pkg/frontend/querymiddleware/codec.go:331]: Stylistic but, I'd just switch on the value of `QueryResultPayloadFormat` here to set the accept header instead of storing a `useX`-style boolean in the struct.

File: pkg/frontend/querymiddleware/codec_test.go [hunk 4]
```diff
@@ -112,6 +112,26 @@ func TestRequest(t *testing.T) {
...
+			encodedRequest, err := codec.EncodeRequest(context.Background(), &req)
+			require.NoError(t, err)
+
+			switch queryResultPayloadFormat {
+			case formatJSON:
+				require.Equal(t, "application/json", encodedRequest.Header.Get("Accept"))
+			case formatProtobuf:
+				require.Equal(t, "application/vnd.mimir.queryresponse+protobuf,application/json", encodedRequest.Header.Get("Accept"))
+			default:
+				panic(fmt.Sprintf("unknown query result payload format: %v", queryResultPayloadFormat))
```
Comment (56quarters) [inline pkg/frontend/querymiddleware/codec_test.go:124]: Seems like this should be `t.Fatal(msg)`, no?

File: pkg/frontend/querymiddleware/codec.go [hunk 5]
```diff
@@ -149,6 +149,16 @@ type prometheusCodec struct {
 	requestProtobufFormat bool
 }
 
+type format interface {
```
Comment (56quarters) [inline pkg/frontend/querymiddleware/codec.go:152]: nit: I'd expect this to be called `formatter`

File: pkg/frontend/querymiddleware/codec.go [hunk 6]
```diff
@@ -149,6 +149,16 @@ type prometheusCodec struct {
...
+	EncodeResponse(resp *PrometheusResponse) ([]byte, error)
+	DecodeResponse([]byte) (PrometheusResponse, error)
```
Comment (56quarters) [inline pkg/frontend/querymiddleware/codec.go:154]: Any reason that this method returns `PrometheusResponse` vs `*PrometheusResponse` here? I don't know that it matters, just seems a little odd since we immediately grab a pointer to it in the `prometheusCodec.DecodeResponse` below.

File: pkg/frontend/querymiddleware/codec.go [hunk 7]
```diff
@@ -349,14 +359,21 @@ func (c prometheusCodec) DecodeResponse(ctx context.Context, r *http.Response, _
 	}
 	log.LogFields(otlog.Int("bytes", len(buf)))
 
+	contentType := r.Header.Get("Content-Type")
+	f, ok := knownFormats[contentType]
```
Comment (56quarters) [inline pkg/frontend/querymiddleware/codec.go:364]: Do we need to make sure we're doing some sort of case normalization here because these are headers?

File: pkg/frontend/querymiddleware/codec_protobuf.go [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+}
+
+func (f protobufFormat) DecodeResponse(buf []byte) (PrometheusResponse, error) {
```
Comment (pracucci) [inline pkg/frontend/querymiddleware/codec_protobuf.go:24]: [thought] It's quite annoying we have to do all this remapping, which I expect will increase allocations significantly. Was this considered when computing the estimated benefit of protobuf?

File: pkg/frontend/querymiddleware/codec_test.go [hunk 9]
```diff
@@ -112,6 +107,26 @@ func TestRequest(t *testing.T) {
 	}
 }
 
+func TestEncodeRequest_AcceptHeader(t *testing.T) {
```
Comment (pracucci) [inline pkg/frontend/querymiddleware/codec_test.go:110]: [nit]
```suggestion
func TestPrometheusCodec_EncodeRequest_AcceptHeader(t *testing.T) {
```

File: pkg/frontend/querymiddleware/codec_test.go [hunk 10]
```diff
@@ -167,235 +182,43 @@ func TestDecodeFailedResponse(t *testing.T) {
...
-func TestResponseRoundtrip(t *testing.T) {
-	headers := http.Header{"Content-Type": []string{"application/json"}}
-	expectedRespHeaders := []*PrometheusResponseHeader{
-		{
-			Name:   "Content-Type",
-			Values: []string{"application/json"},
-		},
-	}
-
+func TestDecodeResponse_ContentTypeHandling(t *testing.T) {
```
Comment (pracucci) [inline pkg/frontend/querymiddleware/codec_test.go:185]: [nit]
```suggestion
func TestPrometheusCodec_DecodeResponse_ContentTypeHandling(t *testing.T) {
```

File: pkg/frontend/querymiddleware/roundtrip_test.go [hunk 11]
```diff
@@ -43,6 +44,7 @@ func TestRangeTripperware(t *testing.T) {
 			http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
 				var err error
 				if r.RequestURI == query {
+					w.Header().Set("Content-Type", jsonMimeType)
```
Comment (pracucci) [inline pkg/frontend/querymiddleware/roundtrip_test.go:47]: Why is this required? I think in case of missing `Content-Type` we should fallback to JSON, so we shouldn't necessarily add this in all tests. Am I missing anything?

File: pkg/frontend/querymiddleware/codec.go [hunk 12]
```diff
@@ -143,12 +145,25 @@ func newPrometheusCodecMetrics(registerer prometheus.Registerer) *prometheusCode
...
+	DecodeResponse([]byte) (PrometheusResponse, error)
+	Name() string
+}
+
+var knownFormats = map[string]format{
+	jsonMimeType:                  jsonFormat{},
+	mimirpb.QueryResponseMimeType: protobufFormat{},
+}
+
+func NewPrometheusCodec(registerer prometheus.Registerer, cfg Config) Codec {
```
Comment (pracucci) [inline pkg/frontend/querymiddleware/codec.go:163]: [design] I suggest to take in input only the preferred response format instead of the whole Config.

**Review by pracucci — APPROVED** [review]
Massive work, good job! I haven't noticed anything critical. I left few comments here and there, mostly minor things.

File: pkg/frontend/querymiddleware/codec_test.go (hunk 4 above)
Comment (charleskorn) [reply pkg/frontend/querymiddleware/codec_test.go]: Fixed in be89dde.

File: pkg/frontend/querymiddleware/codec.go (hunk 6 above)
Comment (charleskorn) [reply pkg/frontend/querymiddleware/codec.go]: Good point - I think that was an artifact of how things were set up pre-refactoring and I didn't think to change it.

Fixed in ea59bdc.

File: pkg/frontend/querymiddleware/codec.go (hunk 7 above)
Comment (charleskorn) [reply pkg/frontend/querymiddleware/codec.go:364]: No - `Header.Get()` will handle this for us:

> It is case insensitive; textproto.CanonicalMIMEHeaderKey is used to canonicalize the provided key.

([source](https://pkg.go.dev/net/http#Header.Get))

File: pkg/frontend/querymiddleware/codec_protobuf_test.go [hunk 13]
```diff
@@ -0,0 +1,477 @@
...
+				Status:    mimirpb.QueryResponse_ERROR,
+				ErrorType: mimirpb.QueryResponse_UNAVAILABLE,
+				Error:     "failed",
+			},
+			expectedErr: apierror.New(apierror.TypeUnavailable, "failed"),
+		},
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			reg := prometheus.NewPedanticRegistry()
+			codec := NewPrometheusCodec(reg, Config{QueryResultPayloadFormat: formatJSON})
```
[the remark this answers is not in the document]
Comment (charleskorn) [reply pkg/frontend/querymiddleware/codec_protobuf_test.go]: Good point, fixed in a7e0908.

File: pkg/frontend/querymiddleware/codec_protobuf.go (hunk 8 above)
Comment (charleskorn) [reply pkg/frontend/querymiddleware/codec_protobuf.go]: Yep, all of my benchmarks included the conversion to `PrometheusResponse`.

File: pkg/frontend/querymiddleware/roundtrip_test.go (hunk 11 above)
Comment (charleskorn) [reply pkg/frontend/querymiddleware/roundtrip_test.go:47]: Queriers (or, more specifically, the Prometheus API code we use) have always set `Content-Type: application/json`, so I favoured making the tests more realistic over adding a fallback that should never be used.

File: pkg/frontend/querymiddleware/codec.go (hunk 12 above)
Comment (charleskorn) [reply pkg/frontend/querymiddleware/codec.go]: Fixed in ed988dc.

File: pkg/frontend/querymiddleware/codec.go [hunk 14]
```diff
@@ -143,12 +145,25 @@ func newPrometheusCodecMetrics(registerer prometheus.Registerer) *prometheusCode
 }
 
 type prometheusCodec struct {
-	metrics *prometheusCodecMetrics
+	metrics               *prometheusCodecMetrics
+	requestProtobufFormat bool
```
[the remark this answers is not in the document]
Comment (charleskorn) [reply pkg/frontend/querymiddleware/codec.go]: Makes sense, changed in cd8674b.

File: pkg/frontend/querymiddleware/codec.go (hunk 3 above)
Comment (charleskorn) [reply pkg/frontend/querymiddleware/codec.go]: Makes sense, changed in cd8674b.

**Comment by pracucci** [comment]
I reviewed it again and I think all feedback has been addressed. I'm going to merge it so it will be included in the next week release.
